config/repository_notification_resetter/notification_resetter/reset_notifications.py

```diff
@@ -0,0 +1,181 @@
+    request = Request(
+        api_url,
+        method="POST",
+        headers={
+            "Authorization": f"Bearer {quay_token}",
+        },
+    )
+    resp: HTTPResponse
+    try:
+        with urlopen(request) as resp:
+            # The actual API response is 204 for notification reset
+            # There is bug in Quay Swagger docs generator
+            # claiming all POST request return 201
+            if resp.status not in (201, 204):
+                raise RuntimeError(resp.reason)
+
+    except HTTPError as ex:
+        # Quay API returns 400 if notification is not found
+        # filter out when this is the case
+        rsp_message = json.loads(ex.read()).get("detail", "")
+        if ex.status == 400 and rsp_message.startswith(
+            "No repository notification found"
+        ):
+            LOGGER.info(
+                "Notification %s from %s/%s was not found", uuid, namespace, name
+            )
+        else:
+            raise (ex)
+
+
+def process_repositories(
+    repos: List[ImageRepo], quay_token: str, dry_run: bool = False
+) -> None:
+    """Process all repositories and reset notifications if needed"""
+    for repo in repos:
+        namespace = repo["namespace"]
+        name = repo["name"]
+        LOGGER.info(
+            "Processing repository %s: %s/%s",
+            next(processed_repos_counter),
+            namespace,
+            name,
+        )
+        all_notifications = get_quay_notifications(quay_token, namespace, name)
+
+        if not all_notifications:
+            continue
+
+        for notification in all_notifications:
+            if notification.get("number_of_failures", 0) > 0:
+                uuid = notification["uuid"]
+                if dry_run:
+                    LOGGER.info(
+                        "Notification %s from %s/%s should be reset",
+                        uuid,
+                        namespace,
+                        name,
+                    )
+                else:
+                    reset_notification(uuid, quay_token, namespace, name)
+                    LOGGER.info(
+                        "Notification %s from %s/%s was reset", uuid, namespace, name
+                    )
+
+
+def fetch_image_repos(access_token: str, namespace: str) -> Iterator[List[ImageRepo]]:
+    """Fetch all image repositories for a given namespace"""
+    next_page = None
+    resp: HTTPResponse
+    while True:
+        query_args = {"namespace": namespace}
+        if next_page is not None:
+            query_args["next_page"] = next_page
+
+        api_url = f"{QUAY_API_URL}/repository?{urlencode(query_args)}"
+        request = Request(
+            api_url,
+            headers={
+                "Authorization": f"Bearer {access_token}",
+            },
+        )
+
+        with urlopen(request) as resp:
+            if resp.status != 200:
+                raise RuntimeError(resp.reason)
```

Comment:
Quay is sometimes unstable and can return weird codes or errors. We already had issues with such a behavior. Does it make sense to just log an error and continue processing other repos? (or maybe retry a few times in case of listing repos). But up to you.